projects/file-parsing/main.go

```diff
@@ -0,0 +1,360 @@
+	// }
+
+	// BIG ENDIAN (starts with fe ff) (00 at the end is the null terminator)
+	// data is displayed in groups of 2 bytes
+	// each group representing a hexadecimal value
+	// byte order where most significant byte is stored first
+
+	// 0000000 [fe ff] [00 00 00 0a] 41 79 61 00 00 00 00 1e 50 72
+	// 0000020 69 73 68 61 00 ff ff ff ff 43 68 61 72 6c 69 65
+	// 0000040 00 00 00 00 19 4d 61 72 67 6f 74 [00]
+	// 0000054
+
+	// LITTLE ENDIAN (starts with ff fe) (00 at the end is the null terminator)
+	// data is displayed in groups of 2 bytes
+	// each group representing a hexadecimal value
+	// byte order where least significant byte is stored first
+
+	// 0000000 [ff fe] [0a 00 00 00] 41 79 61 00 1e 00 00 00 50 72
+	// 0000020 69 73 68 61 00 ff ff ff ff 43 68 61 72 6c 69 65
+	// 0000040 00 19 00 00 00 4d 61 72 67 6f 74 [00]
+	// 0000054
+
+	// initialise two flags (explicitly two, yes could be achieved with one)
+	var bigEndian bool
+	var littleEndian bool
+
+	// check the first two bytes of the file (deliberately using byte() to be explicit)
+	// if they are "fe ff" it is big endian
+	// if they are "ff fe" is is little endian
+	if file[0] == byte(0xfe) && file[1] == byte(0xff) {
+		bigEndian = true
+	} else if file[0] == byte(0xff) && file[1] == byte(0xfe) {
+		littleEndian = true
+	} else {
+		return errors.New("[3] could not establish Endianness of the binary file")
+	}
+
+	// set the start pointer to 2 to skip the first two bytes (which contain the Byte Order Mark BOM)
+	// then loop over the byte slice
+	for startIndex := 2; startIndex < len(file); {
+		// initialise a variable to store the high score
+		var highScore int32
+
+		// parse the high score (signed 32-bit integer) based on the Endianness
+		// (!) but why are we using Uint32 method (when we know its "signed") and then converting it to int32 (presumably "signed").. where is the native xEndian.Int32 method?
+		if bigEndian {
+			highScore = int32(binary.BigEndian.Uint32(file[startIndex : startIndex+4]))
+		} else if littleEndian {
+			highScore = int32(binary.LittleEndian.Uint32(file[startIndex : startIndex+4]))
+		}
+
+		// increment the start pointer by 4
+		startIndex += 4
+
+		// set the end pointer to the start pointer
+		endIndex := startIndex
+
+		// while the value is not a null terminator (deliberately using byte() to be explicit)
+		// walk the end pointer along, to establish the end of the string
+		for file[endIndex] != byte(0x00) {
+			endIndex++
+		}
+
+		// parse the player name (UTF-8)
+		name := string(file[startIndex:endIndex])
+
+		// add the player to the dataSlice
+		*dataSlice = append(*dataSlice, Player{Name: name, HighScore: highScore})
+
+		// adjust the start pointer to skip past the null terminator
+		startIndex = endIndex + 1
+	}
+
+	return nil
+}
+
+func attemptToParse(file []byte) ([]Player, error) {
+	// initialise a slice of Player structs
+	var dataSlice []Player
```

Comment:
There are two major approaches to a function returning a value - either accepting a pointer to a value and populating it, or returning a value.

Can you talk through why you decided `attemptToParseFile` should return a value, whereas each of the `parse*` functions should take pointers to populate?